### PR TITLE
Add CANopen support

### LINKED ISSUE #15278: CANopen Support
**Is your feature request related to a problem? Please describe.**

CANopen is a CAN-based communication system. It comprises higher-layer protocols and profile specifications.

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
This series of commits adds CANopen support to Zephyr through the use of the 3rd party CANopenNode protocol stack.

CANopenNode is under the Apache-2.0 (it was under the GPLv2 license with linking exception when integration begun). It is included as a module in Zephyr.

This fixes #15278.

### CODE REVIEW

**Comment by carlescufi** [comment]
@henrikbrixandersen thanks for the PR. I think you need to submit the license for approval to the TSC, since this is a new license.

**Comment by henrikbrixandersen** [comment]
The checkpatch issues found in `samples/subsys/canbus/canopen/objdict` are due to auto-generated code for the CANopen object dictionary. Would it be okay to exclude this entire folder in `checkpatch.conf`?

**Comment by henrikbrixandersen** [comment]
> 
> 
> @henrikbrixandersen thanks for the PR. I think you need to submit the license for approval to the TSC, since this is a new license.

@carlescufi I have read through the guidelines in https://docs.zephyrproject.org/latest/contribute/index.html#contribute-non-apache but found that they are more relevant for code under a different license that is actually included in the Zephyr Git repository (CANopenNode is included as a module)?

I'll be happy to fill in the format README template for TSC/governing board approval. Do I just put the text in a comment here on the issue and flag it somehow?

**Comment by jukkar** [comment]
> The checkpatch issues found in `samples/subsys/canbus/canopen/objdict` are due to auto-generated code for the CANopen object dictionary.

One option is to run the offending file through uncrustify which hopefully can "fix" the source and then submit the result.

**Comment by henrikbrixandersen** [comment]
> One option is to run the offending file through uncrustify which hopefully can "fix" the source and then submit the result.

@jukkar Thanks for the tip. This would have to be done on every regeneration of the code in question, then (e.g. on all edits of the samples object dictionary). I'll try and see how it goes.

File: subsys/canbus/canopen/CO_driver.c
Comment (alexanderwachter) [inline subsys/canbus/canopen/CO_driver.c]: Are all filters standard frames without mask? If not, the index may increase by more than one and you need more filters. Depending on the filter bank configurations.

File: subsys/canbus/canopen/CO_driver.c
Comment (alexanderwachter) [inline subsys/canbus/canopen/CO_driver.c]: Is there a reason for K_NO_WAIT? Is this called from an ISR?

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

File: samples/subsys/canbus/canopen/src/main.c
Comment (henrikbrixandersen) [reply samples/subsys/canbus/canopen/src/main.c]: No. Updating the object dictionary requires locking (`CO_LOCK_OD()` and `CO_UNLOCK_OD()`).

File: subsys/canbus/canopen/CO_driver.c
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: No, some of the filters are with mask, some without. The number of RX filters needed is determined by the features enabled in the object dictionary (e.g. number of RPDOs) for the given application.

File: subsys/canbus/canopen/CO_driver.c
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: If there are no free TX MBs, the CANopenNode stack expects the API call to return and the CAN frame to be sent at a later time, when a TX MB becomes available (this is handled in the work queue).

Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: `CAN_send()` can be called from "real-time" context, e.g. from the CANopen SYNC thread and must not block.

File: samples/subsys/canbus/canopen/src/main.c
Comment (henrikbrixandersen) [reply samples/subsys/canbus/canopen/src/main.c]: Good catch, thanks. Fixed in three places.

File: samples/subsys/canbus/canopen/src/main.c
Comment (henrikbrixandersen) [reply samples/subsys/canbus/canopen/src/main.c]: The byte ordering, btw, is handled by CANopenNode.

File: subsys/canbus/canopen/CO_driver.c
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: Blocking for as long as it takes to send the CAN frame if a mailbox is available is ok by design.  Blocking while waiting for a free mailbox is not.

CANopenNode does not try to send CAN frames from interrupt context, as far as I can tell. It does call it from high-priority threads, though.

**Comment by henrikbrixandersen** [comment]
> > One option is to run the offending file through uncrustify which hopefully can "fix" the source and then submit the result.
> 
> @jukkar Thanks for the tip. This would have to be done on every regeneration of the code in question, then (e.g. on all edits of the samples object dictionary). I'll try and see how it goes.

Looks like uncrustify isn't sufficient for this. Apart from long lines, C99 comments etc., CO_OD adds typedefs for the complex types used in the object dictionary. It also prefers the types used in CANopenNode over the kernel types preferred in Zephyr (e.g. `u8_t` vs. `uint8_t`).

File: .checkpatch.conf
Comment (jfischer-phytec-iot) [inline .checkpatch.conf]: IMHO file generated by external tool and and not follows coding style belongs to `ext`, there must not be other exception. Also, there are alternatives to this, though not so easy, one is to patch the libedssharp tool and the other to write your own script to generate EDS.

Comment (henrikbrixandersen) [reply .checkpatch.conf]: The generated files will always need to contain typedefs and use the types from CANopenNode, no matter what script generated them.

I can move them to ext, though, if that is preferable.

**Comment by carlescufi** [comment]
> > @henrikbrixandersen thanks for the PR. I think you need to submit the license for approval to the TSC, since this is a new license.
> 
> @carlescufi I have read through the guidelines in https://docs.zephyrproject.org/latest/contribute/index.html#contribute-non-apache but found that they are more relevant for code under a different license that is actually included in the Zephyr Git repository (CANopenNode is included as a module)?
> 
> I'll be happy to fill in the format README template for TSC/governing board approval. Do I just put the text in a comment here on the issue and flag it somehow?

@henrikbrixandersen we quickly discussed this in the TSC yesterday. The guidelines you linked above do apply for modules in the default zephyr upstream manifest, since those are considered part of the zephyr core codebase. Please send an email to: licensereview@lists.zephyrproject.org with the relevant info.

File: samples/subsys/canbus/canopen/README.rst
Comment (dbkinder) [inline samples/subsys/canbus/canopen/README.rst]: Sigh.  You should add after this:
```
Verify that ``~/.local/bin`` is on your PATH so these tools will be found.
```

File: samples/subsys/canbus/canopen/README.rst
Comment (dbkinder) [inline samples/subsys/canbus/canopen/README.rst]: More direct would be to say:
```
To better understand the communication taking place in the following
examples, you can monitor the CAN traffic from the host PC.
```

File: samples/subsys/canbus/canopen/README.rst
Comment (dbkinder) [inline samples/subsys/canbus/canopen/README.rst]: Python (the language) is a proper name and should be capitalized.  Check for more uses of "python" that should be "Python".

**Review by dbkinder — CHANGES_REQUESTED** [review]
a few suggested edits, otherwise docs look good!

File: samples/subsys/canbus/canopen/README.rst
Comment (henrikbrixandersen) [reply samples/subsys/canbus/canopen/README.rst]: Luckily neither python-canopen nor python-can install any binaries so `~/.local/bin` is not required to be in `$PATH` for the following instructions to work.

File: samples/subsys/canbus/canopen/README.rst
Comment (henrikbrixandersen) [reply samples/subsys/canbus/canopen/README.rst]: @dbkinder Thanks, fixed.

File: samples/subsys/canbus/canopen/README.rst
Comment (henrikbrixandersen) [reply samples/subsys/canbus/canopen/README.rst]: @dbkinder Thanks, fixed all over.

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

**Comment by henrikbrixandersen** [comment]
I have been asked a couple of question regarding the exact license of the CANopenNode protocol stack. The correct license identifier, as far as I can tell, is:

SPDX-License-Identifier: GPL-2.0-or-later WITH Classpath-exception-2.0

For reference, see the following pages:
- https://github.com/CANopenNode/CANopenNode/blob/master/CANopen.c (license at the top of the file)
- https://spdx.org/licenses/GPL-2.0-or-later.html
- https://spdx.org/licenses/Classpath-exception-2.0.html

**Comment by henrikbrixandersen** [comment]
@alexanderwachter I have rebased the branch to use your CAN bus state API.

File: subsys/canbus/canopen/CO_driver.c
Comment (alexanderwachter) [inline subsys/canbus/canopen/CO_driver.c]: Don't you want to always enable the mask for RTR?

Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: @alexanderwachter Good catch, thank you! Fixed.

File: subsys/canbus/canopen/CO_driver.c
Comment (alexanderwachter) [inline subsys/canbus/canopen/CO_driver.c]: `err_cnt.tx_err_cnt >= 256U` won't ever evaluate to true. The counter only has 8 bits.
You can use the state return value from `can_get_state` anyway.

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

**Comment by MaureenHelm** [comment]
Thanks @henrikbrixandersen for the contribution. The governing board evaluated the CANopenNode license and decided not to accept it, on the grounds that it is a custom license not approved by the Open Source Initiative (OSI). The board is investigating updating project policy to clearly state that all licenses must be OSI-approved, so future components can be assessed more quickly.

**Comment by henrikbrixandersen** [comment]
CANopenNode has (mostly) been relicensed under the Apache-2.0 license. There are still a few files in the CANopenNode repository under the GNU GPL with Linking Exception license (LSS support and drivers for other platforms than Zephyr). It is expected that the LSS support files will be relicensed shortly and the drivers for other platforms split into separate repositories.

I am reopening this PR to have it reconsidered for being merged into Zephyr.

File: subsys/canbus/canopen/CO_driver.c
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: @alexanderwachter Fixed, thanks.

File: subsys/canbus/canopen/CO_driver.c
Comment (alexanderwachter) [reply subsys/canbus/canopen/CO_driver.c]: I would not do this check. For STM32 you probably need more filters than messages. (because some are with, and some are without a mask.

File: subsys/canbus/canopen/CO_driver.c
Comment (alexanderwachter) [inline subsys/canbus/canopen/CO_driver.c]:
```C
for (i = 0U; i < rxSize; i++) {
	rxArray[i].filter_id = CAN_NO_FREE_FILTER;
}
```
?

File: subsys/canbus/canopen/CO_driver.h
Comment (alexanderwachter) [inline subsys/canbus/canopen/CO_driver.h]: We could probably use typedef struct zcan_frame CO_CANrxMsg_t;
(only a remark)

File: west.yml
Comment (henrikbrixandersen) [reply west.yml]: We need a Zephyr fork to make it into a Zephyr module

File: subsys/canbus/canopen/CO_driver.c
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: This is for initialisation.

File: subsys/canbus/canopen/CO_driver.h
Comment (alexanderwachter) [inline subsys/canbus/canopen/CO_driver.h]: We could embed a zcan_frame. Then we can use the frame directly

File: subsys/canbus/canopen/CO_driver.c
Comment (alexanderwachter) [reply subsys/canbus/canopen/CO_driver.c]: Ah, you are doing this in canopen_detach_all_rx_filters

File: west.yml
Comment (alexanderwachter) [reply west.yml]: @carlescufi can you add a repository for CANopenNode?

File: subsys/canbus/canopen/CO_driver.h
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.h]: No, unfortunately we cannot. The CANopenNode stack expects `struct CO_CANrxMsg_t` to contain at least `data` and `DLC` members (`struct zcan_frame` has `data` and `dlc`).

File: subsys/canbus/canopen/CO_driver.h
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.h]: Same here, the CANopenNode stack expects `struct CO_CANtx_t` to contain struct members with given names.

File: subsys/canbus/canopen/CO_driver.h
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.h]: Fixed.

File: subsys/canbus/canopen/CO_driver.c
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: I would like to keep it in as it makes it much easier to adjust the filter allocation for all other CAN controller drivers.
I have converted the output to `LOG_DBG()` instead.

File: west.yml
Comment (henrikbrixandersen) [reply west.yml]: I have updated currently used repo with a commit to remove all non-Apache-2.0 licensed .h/.c files from the CANopenNode repository as decided on the TSC meeting.

Comment (carlescufi) [reply west.yml]: Done: https://github.com/zephyrproject-rtos/canopennode

Comment (henrikbrixandersen) [reply west.yml]: Thank you. I have completed the import and changed this PR to use the `zephyrproject-rtos/canopennode` repository in `west.yaml`.

**Review by alexanderwachter — APPROVED** [review]

File: samples/subsys/canbus/canopen/README.rst
Comment (jfischer-phytec-iot) [inline samples/subsys/canbus/canopen/README.rst]: Would it be possible to generate the object dictionary at the build time? Or is the dependency on libedssharp not acceptable?

File: subsys/canbus/canopen/CO_driver.c
Comment (henrikbrixandersen) [reply subsys/canbus/canopen/CO_driver.c]: @jfischer-phytec-iot Fixed, thanks.

File: samples/subsys/canbus/canopen/README.rst
Comment (henrikbrixandersen) [reply samples/subsys/canbus/canopen/README.rst]: It is definitely possible to generate the object dictionary implementation files at build time using libedssharp. However, libedssharp is a C#/.NET application/library and, as far as I know, not readily available in binary form (dpkg, rpms, ...).

I am open to adding a dependency on it (since we already use it), but I think it would be a bit much to impose having a C#/.NET development environment setup on all Zephyr users.

Comment (jfischer-phytec-iot) [reply samples/subsys/canbus/canopen/README.rst]: >  am open to adding a dependency on it (since we already use it), but I think it would be a bit much to impose having a C#/.NET development environment setup on all Zephyr users.

That is probably to much :smile:

**Review by jfischer-no — APPROVED** [review]

**Review by chrta — APPROVED** [review]
LGTM, just one minor question/comment.

File: samples/subsys/canbus/canopen/Kconfig
Comment (henrikbrixandersen) [reply samples/subsys/canbus/canopen/Kconfig]: True, but it is just a setting for this sample.

Comment (alexanderwachter) [reply samples/subsys/canbus/canopen/Kconfig]: I'd prefer Kconfig here. It's not a property of the HW/Board which CAN to use for CANopen.

**Review by MaureenHelm — APPROVED** [review]

File: CODEOWNERS
Comment (jukkar) [inline CODEOWNERS]: @henrikbrixandersen Should you place yourself as an owner to files at subsys/canbus/canopen?

**Review by jukkar — APPROVED** [review]
LGTM

File: CODEOWNERS
Comment (alexanderwachter) [reply CODEOWNERS]: @jukkar @henrikbrixandersen asked me to use my github handle for the canbus subsystem.
I think its a good idea not to split the CAN maintainership. We can add him as a collaborator in the new maintainer file then.

**Pull request merged**
